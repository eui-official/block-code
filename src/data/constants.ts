import { TInputProps } from '../types'

const EMPTY_VALUE = ''

const DEFAULT_INPUT_PROPS: TInputProps = {
  maxLength: 1,
  className: 'input'
}

export { EMPTY_VALUE, DEFAULT_INPUT_PROPS }
