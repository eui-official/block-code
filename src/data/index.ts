export * from './constants'
export * from './keyboard'
