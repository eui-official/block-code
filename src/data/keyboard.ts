const DELETE = 'Delete'
const BACK_SPACE = 'Backspace'
const ARROW_LEFT = 'ArrowLeft'
const ARROW_RIGHT = 'ArrowRight'

export const KEYLIST = {
  DELETE,
  BACK_SPACE,
  ARROW_LEFT,
  ARROW_RIGHT
}
