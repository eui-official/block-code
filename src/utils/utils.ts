export const Utils = {
  findIndex: (
    index: number,
    length: number,
    maxLength: number,
    maxInputs: number
  ) => {
    if (length === maxLength && index < maxInputs) {
      ++index
    } else if (length === 0 && index > 0) {
      --index
    }

    return index
  },

  createId: (prefix = 'block-code') =>
    `${prefix}-${Math.random().toString(36).substring(2, 9)}`,

  createChunk: (value: string, length: number) =>
    value.match(new RegExp('.{1,' + length + '}', 'g'))
}
