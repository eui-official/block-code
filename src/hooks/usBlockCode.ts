import { DEFAULT_INPUT_PROPS, EMPTY_VALUE, KEYLIST } from '../data'
import {
  ClipboardEvent,
  FormEvent,
  KeyboardEvent,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from 'react'
import {
  IOnValidateBeforeChange,
  TDataSource,
  TInputFormEvent,
  TInputProps,
  TOnComplete,
  TOnChange,
  TOnValidateBeforeChange
} from '../types'
import { Utils } from '../utils'

export interface IUseBlockCodeProps {
  /**
   * Call when values got changed
   */
  onValuesChange?: TOnChange<string[]>
  /**
   * Call when onInput and onKeyDown
   */
  onInputChange?: TOnChange<TInputFormEvent>
  /**
   * Validate value before changing
   */
  onValidateBeforeChange?: TOnValidateBeforeChange<string, TDataSource>
  /**
   * Only call when enough data
   */
  onComplete?: TOnComplete<string>
  /**
   * How many input we're gonna show
   * @default 6
   */
  maxInputs?: number
  /**
   * Input props pass to input element
   */
  inputProps?: TInputProps
  /**
   * Auto focus to the first input
   * @default true
   */
  autoFocus?: boolean
}

export const useBlockCode = (props: IUseBlockCodeProps = {}) => {
  const {
    onValuesChange,
    onValidateBeforeChange,
    onInputChange,
    onComplete,
    maxInputs = 6,
    inputProps,
    autoFocus = true
  } = props

  const InputProps = Object.assign({}, DEFAULT_INPUT_PROPS, inputProps)

  /**
   * Declare default values
   */
  const ids = useMemo(
    () => [...Array(maxInputs)].map(Utils.createId),
    [maxInputs]
  )
  const initState = useMemo(
    () => Array<string>(maxInputs).fill(EMPTY_VALUE),
    [maxInputs]
  )
  const initActiveIndex = useMemo(() => (autoFocus ? 0 : -1), [autoFocus])

  /**
   * Declare states
   */
  const [activeIndex, setActiveIndex] = useState<number>(initActiveIndex)
  const [values, setValues] = useState(initState)

  const refs = useRef<HTMLInputElement[]>([])

  /**
   * Reset values to initState which is empty
   */
  const onResetValues = useCallback(() => {
    setValues(initState)
  }, [initState])

  /**
   * Reset the state. It will reset:
   *
   * @values to initState
   * @index to 0
   */
  const onReset = useCallback(() => {
    onResetValues()
    setActiveIndex(0)
  }, [onResetValues])

  /**
   * Check if data is enough then call `onComplete`
   */
  const onEnoughData = useCallback(() => {
    const { maxLength } = InputProps

    if (maxLength) {
      const raw = values.filter(data => data)
      const value = raw.join(EMPTY_VALUE)
      const isEnoughData = value.length == maxInputs * maxLength

      if (isEnoughData && onComplete) {
        onComplete({ value, raw, onReset })
      }
    }
  }, [InputProps, maxInputs, onComplete, onReset, values])

  /**
   * Check the data if valid
   *
   * @param data Value which wants to be added into the state
   * @param onCallback Callback will be called after passing conditions
   */
  const onCheckValue = async (
    data: IOnValidateBeforeChange,
    onCallback: VoidFunction
  ) => {
    if (onValidateBeforeChange) {
      const isValid = await onValidateBeforeChange(data)

      isValid && onCallback()
    } else {
      onCallback()
    }
  }

  /**
   *  Add a new data to the state, however it must fits the condition from `onValidateBeforeChange`
   *
   * @param data Data which wants to be added into the state
   * @param forIndex The index we want to set value for
   * @param onCallback Callback will be called after adding value
   */
  const onAddValue = (
    data: IOnValidateBeforeChange,
    forIndex: number,
    onCallback?: VoidFunction
  ) => {
    onCheckValue(data, () => {
      onSetValueAtIndex(data.value, forIndex)
      onCallback && onCallback()
    })
  }

  /**
   *  Trigger `onInputChange` to the user
   *
   * @param event The event which makes values got changed
   */
  const onAfterAddingValue = (event: TInputFormEvent) => {
    onInputChange && onInputChange(event)
  }

  /**
   *  Add ref to the refs array so we can control later
   *
   * @param index Index of the input
   * @param ref Ref of the input
   */
  const onCreateRef = (index: number, ref: HTMLInputElement) => {
    refs.current[index] = ref
  }

  /**
   * Update state values
   * @param value A new value of the input
   * @param forIndex The index we want to set value for
   */
  const onSetValueAtIndex = (value: string, forIndex: number) => {
    setValues(prevState => {
      const newState = [...prevState]
      newState[forIndex] = value
      return newState
    })
  }

  /**
   * Check if the index is valid, not smaller than 0 and greater than `maxInputs`
   *
   * If fits the condition then call `setActiveIndex`
   */
  const onSetActiveIndex = useCallback(
    (index: number) => {
      const nextActiveIndex = Math.max(Math.min(maxInputs - 1, index), 0)
      setActiveIndex(nextActiveIndex)
    },
    [maxInputs]
  )

  /**
   * Focus to the next input
   */
  const onNextFocusAfterChangingValue = useCallback(
    (value: string) => {
      const { maxLength } = InputProps
      const { length } = value

      if (maxLength) {
        const index = Utils.findIndex(activeIndex, length, maxLength, maxInputs)
        onSetActiveIndex(index)
      }
    },
    [InputProps, activeIndex, maxInputs, onSetActiveIndex]
  )

  /**
   * Detect the user change values
   * @param event FormEvent
   */
  const onInput: TOnChange<FormEvent<HTMLInputElement>> = event => {
    const { value } = event.currentTarget

    onAddValue({ value, event }, activeIndex, () => {
      onNextFocusAfterChangingValue(value)
      onAfterAddingValue(event)
    })
  }

  /**
   * Detect the user paste values into the input
   * @param event ClipboardEvent
   */
  const onPaste: TOnChange<ClipboardEvent<HTMLInputElement>> = async event => {
    event.preventDefault()
    const { maxLength } = InputProps

    if (maxLength) {
      const originaltext = await navigator.clipboard.readText()
      const trim = originaltext.trim()
      const chunk = Utils.createChunk(trim, maxLength)

      if (chunk) {
        let index = activeIndex

        for (let i = activeIndex; i < maxInputs; i++) {
          const value = chunk.shift()

          if (value) {
            onAddValue({ value, event }, i)
            index = i
          }
        }

        onSetActiveIndex(index)
      }
    }
  }

  /**
   * Detect the user press a key
   * @param event KeyboardEvent
   */
  const onKeyDown: TOnChange<KeyboardEvent<HTMLInputElement>> = event => {
    let index = activeIndex
    const { key } = event
    const isInKeyList = Object.values(KEYLIST).includes(key)
    const isDelete = [KEYLIST.BACK_SPACE, KEYLIST.DELETE].includes(key)

    if (isInKeyList) event.preventDefault()

    if (isDelete) {
      onSetValueAtIndex(EMPTY_VALUE, activeIndex)
      onAfterAddingValue(event)
    }

    switch (key) {
      case KEYLIST.BACK_SPACE:
      case KEYLIST.DELETE:
      case KEYLIST.ARROW_LEFT:
        --index
        break
      case KEYLIST.ARROW_RIGHT:
        ++index
        break
    }

    if (isInKeyList) onSetActiveIndex(index)
  }

  /**
   *  A helper function to generate input props
   *
   * @param index Index of the input
   * @returns Input props
   */
  const onCreateInputProps = (index: number): TInputProps => {
    return {
      value: values[index],
      ref: ref => ref && onCreateRef(index, ref),
      onFocus: () => onSetActiveIndex(index),
      onKeyDown,
      onInput,
      onPaste,
      ...InputProps
    }
  }

  /**
   * When values got changed, we need to check if the data is enough?
   * If it's enough and fits all the condition
   * We'll call `onComplete` event
   */
  useEffect(() => {
    onEnoughData()
  }, [onEnoughData, values])

  /**
   * When `values` got changed, send it to the user
   */
  useEffect(() => {
    onValuesChange?.(values)
  }, [onValuesChange, values])

  /**
   * When `activeIndex` got changed, check and focus to the next input
   */
  useEffect(() => {
    const element = refs.current[activeIndex]

    if (element) {
      element.focus()
      element.select()
    }
  }, [activeIndex])

  return {
    values,
    setValues,
    setActiveIndex,
    refs,
    onSetValueAtIndex,
    onSetActiveIndex,
    onResetValues,
    onReset,
    onPaste,
    onNextFocusAfterChangingValue,
    onKeyDown,
    onInput,
    onEnoughData,
    onCreateRef,
    onCreateInputProps,
    onCheckValue,
    onAfterAddingValue,
    onAddValue,
    initState,
    initActiveIndex,
    ids,
    activeIndex
  }
}
