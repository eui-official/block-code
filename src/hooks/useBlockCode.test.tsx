import { fireEvent, render } from '@testing-library/react'
import { renderHook } from '@testing-library/react-hooks'
import { describe, expect, test, vi } from 'vitest'
import { EMPTY_VALUE, KEYLIST } from '../data'
import { useBlockCode } from './usBlockCode'
import { Utils } from '../utils'

const DEFAULT_VALUES = ['1', '2', '3', '4', '5', '6']

const REVERSE_DEFAULT_VALUES = [...DEFAULT_VALUES].reverse()

describe('useBlockCode', () => {
  test('Should return input props correctly', () => {
    const spyCreateId = vi.spyOn(Utils, 'createId')
    const { result } = renderHook(() => useBlockCode())
    const { ids, ...restProps } = result.current

    expect(spyCreateId).toBeDefined()
    expect(spyCreateId).toHaveBeenCalledTimes(6)
    expect(ids.length).toBe(6)
    expect(restProps).toMatchSnapshot()
  })

  test('Should call setValues correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    expect(result.current.values).toMatchSnapshot()

    result.current.setValues(DEFAULT_VALUES)
    expect(result.current.values).toMatchSnapshot()

    result.current.setValues(REVERSE_DEFAULT_VALUES)
    expect(result.current.values).toMatchSnapshot()

    result.current.setValues([])
    expect(result.current.values).toMatchSnapshot()
  })

  test('Should call setActiveIndex correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    expect(result.current.activeIndex).toBe(0)
    result.current.setActiveIndex(1)
    expect(result.current.activeIndex).toBe(1)
  })

  test('Should call setActiveIndex correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    expect(result.current.activeIndex).toBe(0)
    result.current.setActiveIndex(1)
    expect(result.current.activeIndex).toBe(1)
  })

  test('Should call onSetValueAtIndex correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    expect(result.current.values[0]).toBe(EMPTY_VALUE)
    result.current.onSetValueAtIndex('1', 0)
    expect(result.current.values[0]).toBe('1')
  })

  test('Should call nSetActiveIndex correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    expect(result.current.activeIndex).toMatchSnapshot()
    result.current.onSetActiveIndex(1)
    expect(result.current.activeIndex).toMatchSnapshot()
  })

  test('Should call onResetValues correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    expect(result.current.values).toMatchSnapshot()

    result.current.setValues(DEFAULT_VALUES)
    expect(result.current.values).toMatchSnapshot()

    result.current.onResetValues()

    expect(result.current.values).toMatchSnapshot()
  })

  test('Should call onReset correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    expect(result.current.values).toMatchSnapshot()
    expect(result.current.activeIndex).toMatchSnapshot()

    result.current.onSetValueAtIndex('1', 2)
    result.current.onSetActiveIndex(2)
    expect(result.current.values[2]).toMatchSnapshot()
    expect(result.current.activeIndex).toMatchSnapshot()

    result.current.onReset()

    expect(result.current.values).toMatchSnapshot()
    expect(result.current.activeIndex).toMatchSnapshot(0)
  })

  test('Should call onEnoughData correctly', () => {
    const mockOnComplete = vi.fn()
    const { result } = renderHook(() =>
      useBlockCode({
        onComplete: mockOnComplete
      })
    )

    result.current.setValues(DEFAULT_VALUES)
    result.current.onEnoughData()
    expect(mockOnComplete).toHaveBeenCalled()
  })

  test('Should not call onEnoughData if the data is not enough', () => {
    const mockOnComplete = vi.fn()
    const { result } = renderHook(() =>
      useBlockCode({
        onComplete: mockOnComplete
      })
    )

    result.current.onEnoughData()
    expect(mockOnComplete).not.toHaveBeenCalled()
  })

  test('Should call onCreateInputProps correctly', () => {
    const { result } = renderHook(() => useBlockCode())
    expect(result.current.onCreateInputProps(0)).toMatchSnapshot()
  })

  test('Should call onValuesChange correctly', async () => {
    const spyOnValuesChange = vi.fn()
    const { result, waitFor } = renderHook(() =>
      useBlockCode({ onValuesChange: spyOnValuesChange })
    )

    const { getByLabelText } = render(
      <input
        aria-label="block-code-values-change"
        onInput={result.current.onInput}
      />
    )
    const input = getByLabelText('block-code-values-change')
    fireEvent.input(input, { currentTarget: { value: 1 } })

    await waitFor(() => {
      expect(result.current.values).toMatchSnapshot()
      expect(spyOnValuesChange).toHaveBeenCalled()
    })
  })

  test('Should call onValuesChange correctly', async () => {
    const spyOnInputChange = vi.fn()
    const { result, waitFor } = renderHook(() =>
      useBlockCode({ onInputChange: spyOnInputChange })
    )

    const { getByLabelText } = render(
      <input
        aria-label="block-code-input-change"
        {...result.current.onCreateInputProps(0)}
      />
    )
    const input = getByLabelText('block-code-input-change')
    fireEvent.input(input, { currentTarget: { value: 1 } })

    await waitFor(() => {
      expect(result.current.values).toMatchSnapshot()
      expect(spyOnInputChange).toHaveBeenCalled()
    })
  })

  test('Should call onValidateBeforeChange correctly', async () => {
    const spyOnValidateBeforeChange = vi.fn()
    const { result, waitFor } = renderHook(() =>
      useBlockCode({ onValidateBeforeChange: spyOnValidateBeforeChange })
    )

    const { getByLabelText } = render(
      <input
        aria-label="block-code-validate-before-on-change"
        {...result.current.onCreateInputProps(0)}
      />
    )
    const input = getByLabelText('block-code-validate-before-on-change')
    fireEvent.input(input, { currentTarget: { value: 1 } })

    await waitFor(() => {
      expect(result.current.values).toMatchSnapshot()
      expect(spyOnValidateBeforeChange).toHaveBeenCalled()
    })
  })

  test('Should call onCreateRef correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    const { getByLabelText } = render(
      <input
        aria-label="block-code-create-ref"
        {...result.current.onCreateInputProps(0)}
      />
    )
    const input = getByLabelText('block-code-create-ref') as HTMLInputElement

    expect(result.current.refs.current[0]).toBe(input)
  })

  test('Should call onPaste correctly', () => {
    const { result } = renderHook(() =>
      useBlockCode({
        onValidateBeforeChange: async ({ value }) => {
          expect(DEFAULT_VALUES.includes(value)).toBeTruthy()
          return true
        }
      })
    )

    const { getByLabelText } = render(
      <input
        aria-label="block-code-on-paste"
        {...result.current.onCreateInputProps(0)}
      />
    )

    const input = getByLabelText('block-code-on-paste')
    fireEvent.paste(input)

    /**
     * TODO:
     * - Test the state propertly when vitest supports state changes
     */
  })

  test('Should call onKeyDown correctly', () => {
    const { result } = renderHook(() => useBlockCode())

    const { getByLabelText } = render(
      <input
        aria-label="block-code-on-key-down"
        {...result.current.onCreateInputProps(0)}
      />
    )

    const input = getByLabelText('block-code-on-key-down')

    fireEvent.keyDown(input, { key: KEYLIST.ARROW_LEFT })
    fireEvent.keyDown(input, { key: KEYLIST.ARROW_RIGHT })
    fireEvent.keyDown(input, { key: KEYLIST.BACK_SPACE })
    fireEvent.keyDown(input, { key: KEYLIST.DELETE })

    /**
     * TODO:
     * - Test the state propertly when vitest supports state changes
     */
  })
})
