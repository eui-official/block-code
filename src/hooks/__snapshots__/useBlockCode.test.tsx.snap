// Vitest Snapshot v1

exports[`useBlockCode > Should call nSetActiveIndex correctly 1`] = `0`;

exports[`useBlockCode > Should call nSetActiveIndex correctly 2`] = `1`;

exports[`useBlockCode > Should call onCreateInputProps correctly 1`] = `
{
  "className": "input",
  "maxLength": 1,
  "onFocus": [Function],
  "onInput": [Function],
  "onKeyDown": [Function],
  "onPaste": [Function],
  "ref": [Function],
  "value": "",
}
`;

exports[`useBlockCode > Should call onReset correctly 1`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`useBlockCode > Should call onReset correctly 2`] = `0`;

exports[`useBlockCode > Should call onReset correctly 3`] = `"1"`;

exports[`useBlockCode > Should call onReset correctly 4`] = `2`;

exports[`useBlockCode > Should call onReset correctly 5`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`useBlockCode > Should call onReset correctly 6`] = `0`;

exports[`useBlockCode > Should call onResetValues correctly 1`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`useBlockCode > Should call onResetValues correctly 2`] = `
[
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
]
`;

exports[`useBlockCode > Should call onResetValues correctly 3`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`useBlockCode > Should call onValidateBeforeChange correctly 1`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`useBlockCode > Should call onValuesChange correctly 1`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`useBlockCode > Should call onValuesChange correctly 2`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`useBlockCode > Should call setValues correctly 1`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`useBlockCode > Should call setValues correctly 2`] = `
[
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
]
`;

exports[`useBlockCode > Should call setValues correctly 3`] = `
[
  "6",
  "5",
  "4",
  "3",
  "2",
  "1",
]
`;

exports[`useBlockCode > Should call setValues correctly 4`] = `[]`;

exports[`useBlockCode > Should return input props correctly 1`] = `
{
  "activeIndex": 0,
  "initActiveIndex": 0,
  "initState": [
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  "onAddValue": [Function],
  "onAfterAddingValue": [Function],
  "onCheckValue": [Function],
  "onCreateInputProps": [Function],
  "onCreateRef": [Function],
  "onEnoughData": [Function],
  "onInput": [Function],
  "onKeyDown": [Function],
  "onNextFocusAfterChangingValue": [Function],
  "onPaste": [Function],
  "onReset": [Function],
  "onResetValues": [Function],
  "onSetActiveIndex": [Function],
  "onSetValueAtIndex": [Function],
  "refs": {
    "current": [],
  },
  "setActiveIndex": [Function],
  "setValues": [Function],
  "values": [
    "",
    "",
    "",
    "",
    "",
    "",
  ],
}
`;
